### PR TITLE
feat(Ida.sol): add functions for validator

### PR DESCRIPTION
**Pull request by unknownunknown1**

### CODE REVIEW

File: contracts/Ida.sol [hunk 1]
```diff
@@ -84,22 +88,40 @@ contract Ida {
         emit Funded(msg.sender, _amount, impactPromise.totalSupply());
     }
 
+    function registerReport(bytes32 key) public onlyValidator {
+      require(!hasEnded(), "Cannot register after project end");
+      require(!reportRegistered[key], "The report is already registered!");
+      reportRegistered[key] = true;
+      require(claimsRegistry.getClaim(serviceProvider, address(this), key) == bytes32(promisePrice), "A claim must be registered before registering a report");
```
Comment (jakub-wojciechowski) [inline contracts/Ida.sol:95]: Shouldn't we also check:
require(!claimsRegistry.isApproved(validator, serviceProvider, address(this), key), "This promise has already been validated");
to avoid reporting for already validated promises ?

File: contracts/Ida.sol [hunk 2]
```diff
@@ -84,22 +88,40 @@ contract Ida {
...
+      nbPending = nbPending.add(1);
+    }
 
     function validatePromise(bytes32 key) public onlyValidator {
-        require(!hasEnded(), "Cannot validate after project end");
-        require(validatedNumber < promiseNumber, "All the promises have already been validated");
```
Comment (jakub-wojciechowski) [inline contracts/Ida.sol:90]: Do you mind asking why have you converted the "require(validatedNumber < promiseNumber, "All the promises have already been validated");" precondition to if statement?

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: contracts/Ida.sol (hunk 1 above)
Comment (unknownunknown1) [reply contracts/Ida.sol:95]: I didn't put it there because on validator's side we only allow to register new reports. `require(report.status == Status.None, "The report for this impact promise has already been created.");`
So if the report has been created and even approved, it wouldn't be possible to translate it to Ida.
But I agree that this check still can be added for readability and consistency, and also for extra security.

File: contracts/Ida.sol (hunk 2 above)
Comment (unknownunknown1) [reply contracts/Ida.sol:90]: On telegram channel I mentioned a problem that if we count the number of pending reports (and close Ida if this number reaches 0 at the end), we should make sure that each report will be fully processed and not reverted at the end of the lifecycle.
In order to do so, each check we do at the end I also put at the very beginning so we would know from the start the report is legit and won't be reverted at the end and the nbPending will be decreased.
However, I don't think there is a way to check that the report exceeds the required promiseNumber at the moment when it's created.
Let's say, validatedNumber is 0 and promiseNumber is 2. We register 3 reports at the same time (some of them might be legit, some not), and turns out all 3 of them get validated. At this point the last report wouldn't be processed fully because of this require, so I had to soften it to simple condition, so validate function doesn't revert in this case and nbPending is successfully decreased.
I hope it's all clear :)
There still might be a solution to this without changing the require though

Comment (jakub-wojciechowski) [reply contracts/Ida.sol:90]: Hmm, I'm afraid I'm still don't fully get it. It'll be helpful for me to see the multiple report scenario (3 reports for 2 promises), especially as you've added the validation: require(!reportRegistered[key], "The report is already registered!"); in registerReport function?

Comment (unknownunknown1) [reply contracts/Ida.sol:90]: This require prevents from submitting the same report multiple times, but I'm thinking the situation where the reports are made for different promises and the number of reports exceeds promiseNumber set in Ida. Or is this scenario is not at all possible?

Comment (jakub-wojciechowski) [reply contracts/Ida.sol:90]: That was my question :) I think your validation prevents such a scenario from happening but I wanted to double check if you see any possible way to create more reports than promises?

Comment (unknownunknown1) [reply contracts/Ida.sol:90]: Oh, ok :)
Initially I assumed it's possible and this assumption was based mainly on this require
`require(validatedNumber < promiseNumber, "All the promises have already been validated");`
In practice, however, I'm not sure. I think it can happen if service provider goes directly to ClaimsRegistry and invokes setClaim for excessive number of promise keys. Like, Ida only expects to have 2 promises with keys "a" and "b" and provider invokes setClaim for keys "a", "b", "c" and "d". Then it'd be possible to pass this require two more times
`require(claimsRegistry.getClaim(serviceProvider, address(this), key) == bytes32(promisePrice), "A claim must be registered before registering a report");`
and have too many legit reports.

Comment (jakub-wojciechowski) [reply contracts/Ida.sol:90]: That's the scenario I've been looking for, thanks for catching it. Ideally, we should prevent it in the registerReport method to check precondition if (validatedNumber +nbPending < promiseNumber). First we don't expect more reports than promises so it's a sign of an error, second we can avoid an unnecessary race condition where there are more valid reports than promise slots. What do you think about it? 
It also pointed me to another potential vulnerability. A malicious service provider may update an entry in claims registry changing the value after the report was submitted. Then we will have a phantom report that couldn't be validated and will block the Ida. Do you also see a problem there?
If you agree with the suggestions feel free to update the code. I'm also happy to merge it and update afterwards myself as I don't want to over-utilise your generosity and effort for which we are extremely grateful :)

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: contracts/Ida.sol (hunk 2 above)
Comment (unknownunknown1) [reply contracts/Ida.sol:90]: >to check precondition if (validatedNumber +nbPending < promiseNumber).

This will work, however I'd like to point out that a pending report won't necessarily be validated at the end so it doesn't mean it will take a slot. But it's not necessarily an issue and probably more like an edge case.

> A malicious service provider may update an entry in claims registry changing the value after the report was submitted.

I'm not sure that's a huge problem considering that service provider is basically the owner of Ida so if he wants to sabotage his own project it's his choice, I guess. This particular example seems reversible though because the entry in registry can be changed back to be valid again. But regardless, we can  easily solve this potential issue by removing "getClaim" check in validatePromise function.
So to summarize: 
1) We need to change the condition for validatedNumber back to require and move it to registerReport function.
2) Remove getClaim check from validatePromise to prevent locking Ida by malicious provider.
3) Add "isApproved" check to registerReport as an extra security measure

Please let me know if I got it correct. I can update the code myself :)
Also, Ida.sol and Validator.sol haven't been reviewed yet by our team, so it still can have some potential issues.

Comment (jakub-wojciechowski) [reply contracts/Ida.sol:90]: Thanks for looking into it. The list of proposed changes looks really good. We're going to have an external audit after we integrate with you guys and incorporate some feedback from users.

Comment (unknownunknown1) [reply contracts/Ida.sol:90]: I updated the PR. Please, verify

**Review by jwojciechowski — APPROVED** [review]

**Comment by jakub-wojciechowski** [comment]
Thanks for the PR @unknownunknown1 !!!
